Bugfix-1837: Metacat Admin Privileges Extended

doulikecookiedough opened the pull request:
Greetings @artntek 

I have moved the logic to check for Metacat admin privileges from the `doAdminAuthorization` method to `isLocalNodeAdmin`.

The `isLocalNodeAdmin` is called by the 6 public methods defined in the javadoc used to check for authorization, so it looks like we should have adequate coverage now.

I have also added additional junit tests to check for Metacat admin privileges and cleaned up the test class (easier to follow now). Please let me know if you have any feedback. 

Note - We are missing an accurate test for the `expandRightsHolder` method. Can you please help  take a look?

**This PR addresses the following issue:** [Metacat admin privileges are not checked for all methods used for authorization](https://github.com/NCEAS/metacat/issues/1837)

doulikecookiedough commented:
@artntek @taojing2002 This PR is now ready for your review. Thank you!

artntek commented on test/edu/ucsb/nceas/metacat/dataone/D1AuthHelperTest.java:
```diff
@@ -157,44 +182,293 @@ public void setUp() throws Exception {
+        // D1Client contains methods that make network calls, which is to be mocked
+        try (MockedStatic<D1Client> mockD1Client = Mockito.mockStatic(D1Client.class)) {
+            // Create a member list, to be added to the group
+            List<Subject> hasMemberList = new ArrayList<>();
+            Subject testGroupMember = new Subject();
+            testGroupMember.setValue("testGroupMember");
+            hasMemberList.add(testGroupMember);        // Bogus value
+            hasMemberList.add(sysmeta.getSubmitter()); // The matching member
+            rightsHolderGroup.setHasMemberList(hasMemberList);
+
+            SubjectInfo mockSInfo = Mockito.mock(SubjectInfo.class);
+            List<Group> groups = new ArrayList<>();
+            groups.add(rightsHolderGroup);
+            when(mockSInfo.getGroupList()).thenReturn(groups);
+
+            CNode mockCN = Mockito.mock(CNode.class);
```
these lines are repeated verbatim in 2 other tests. Can we extract into a private method?

(bonus points for making it work with the 4th method, too ;-) )

doulikecookiedough replied:
Extracted, and refactored the relevant junit tests.

artntek commented on test/edu/ucsb/nceas/metacat/dataone/D1AuthHelperTest.java:
```diff
@@ -157,44 +182,293 @@ public void setUp() throws Exception {
+    }
+
+    /**
+     * Confirm that expandRightsHolder should return false when the suppled rightsHolder
```
typo "supplied" (unless you mean the rights holder does a lot of stretching)

artntek commented on test/edu/ucsb/nceas/metacat/dataone/D1AuthHelperTest.java:
```diff
@@ -83,12 +91,21 @@ public static void setUpBeforeClass() {
 
     D1AuthHelper authDel;
     D1AuthHelper authDelMock;
-    Session session;
+    Session defaultSession;
     Session authMNSession;
     Session otherMNSession;
     Session replMNSession;
     Session cn1CNSession;
+    Session nullSession;
+    Session missingSubjectSession;
+    Session emptySubjectSession;
+    Session metacatAdminSession;
+    Session metacatAdminOtherSession;
+    Session localNodeSession;
+    Session notAuthorizedSession;
     SystemMetadata sysmeta;
+    Group rightsHolderGroup;
+    Subject rhgSubject;
```
can these go before line 45?

artntek commented on test/edu/ucsb/nceas/metacat/dataone/D1AuthHelperTest.java:
```diff
@@ -97,7 +114,8 @@ private SystemMetadata getGenericSysmetaObject() throws Exception {
             TypeFactory.buildMinimalSystemMetadata(TypeFactory.buildIdentifier("dip"),
                                                    new ByteArrayInputStream(
-                                                       ("tra la la la la").getBytes("UTF-8")),
+                                                       ("tra la la la la").getBytes(
```
it's good practice to give some indication of what this is, instead of a totally unrelated content. When debugging, it's easier to figure out where "test Sysmeta data inputstream" came from, versus "tra la la la la". (great tune, though)

doulikecookiedough replied:
Understood, thank you and I've updated the string content.

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
      */
     protected boolean isLocalNodeAdmin(Session session, NodeType nodeType) throws ServiceFailure {
-
         boolean allowed = false;
```
I find it harder to follow the logic of methods like this, that initialize a boolean, then assign different values to it in different blocks, and finally return it at the end. I think it's clearer and easier to follow, if the return value is simply returned immediately instead of assigned for returning later, so you can tell by looking at the code, exactly what the return value is. I have annotated the relevant lines in the code below, to show what I mean:

artntek replied:
first, delete line 730; then...

doulikecookiedough replied:
Thank you for outlining how to refactor/improve this method. I have made the changes, and it is indeed a lot easier to follow/understand now.

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
+        try {
+            sessionSubjectValue = checkSessionAndGetSubjectValue(session);
+        } catch (NotAuthorized na) {
+            return allowed;
```
`return false;`

taojing2002 commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
Please removed the unused import files.

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
+                            + nodeSubject.getValue() + " and the session user: "
+                            + subject.getValue());
+                        if (nodeSubject.equals(subject)) {
+                            allowed = true; // subject of session == this node's subject
                             break outer;
```
delete these 2 lines. replace with:

```java
    logMetacat.debug("Found " + subject.getValue() + " in the node subject list. Returning true");
    return true;
```

(no need to include D1AuthHelper.isLocalNodeAdmin() - this is already added to the end of the log line by log4j)

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
+                            + " for Metacat admin privileges.");
+                        try {
+                            if (AuthUtil.isAdministrator(subjectValue, null)) {
+                                allowed = true;
+                                break outer;
```
delete these 2 lines. replace with:

```java
    logMetacat.debug("Found " + subjectValue + " in the metacat admins list. Returning true");
    return true;
```

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
             List<Subject> nodeSubjects = node.getSubjectList();
             if (sessionSubjects != null) {
-              outer:
+                outer:
```
delete this line

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
         }
-        logMetacat.debug("In is isLocalNodeAdmin method. Is this a local node admin? "+allowed);
+
+        logMetacat.debug(
+            "D1AuthHelper.isLocalNodeAdmin method. Is this a local node admin? " + allowed);
         return allowed;
```
delete and replace with:

```java
    logMetacat.debug("Not a local node admin; returning false");
    return false;
```

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
+                                break outer;
+                            }
+                        } catch (MetacatUtilException mue) {
+                            throw new ServiceFailure("0000", mue.getMessage());
```
might be better not to throw an exception here, and instead allow the outer loop to complete, since it may find matches in subsequent loops?

doulikecookiedough replied:
This makes sense, nice catch!

artntek commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
lots of linting errors that could be fixed (unused imports etc). Let's leave things better than we found them :-)

doulikecookiedough replied:
Got it! I will keep that in mind next time. Thank you for the reminder and nudge.

doulikecookiedough commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -726,54 +715,72 @@ public boolean isLocalCNAdmin(Session session) throws ServiceFailure {
+                    }
+                    // If not, check session subject for a metacat admin
+                    String subjectValue = subject.getValue();
+                    if (subjectValue != null || !subjectValue.trim().isBlank()) {
```
Thank you! This has been resolved.

doulikecookiedough commented on src/edu/ucsb/nceas/metacat/dataone/D1AuthHelper.java:
```diff
@@ -153,6 +153,7 @@ public void doIsAuthorized(Session session, SystemMetadata sysmeta, Permission p
 
         // this makes 1 or more calls to listSubjects, so is the most expensive
         try {
+            logMetacat.debug("D1AuthHelper.doIsAuthorized - Checking expanded permissions");
```
I have moved the debug statement.

doulikecookiedough commented:
Thank you for the detailed and helpful feedback @artntek. I believe I have addressed all the issues raised, I've also cleaned up `D1AuthHelper` to eliminate most of the linting errors and warnings (missing javadocs, IDE suggested changes, etc.). Note - there are a lot of formatting changes to fix line lengths and spacing. Please let me know if there's anything else you think I should take a look at.

artntek commented on test/edu/ucsb/nceas/metacat/dataone/D1AuthHelperTest.java:
```diff
@@ -81,33 +106,44 @@ public static void setUpBeforeClass() {
+    /**
+     * Get a rightsHolderGroup, if null is supplied default values will be used.
+     */
+    private List<Group> getGenericRightsHolderGroup(List<Subject> subjectMemberList) {
```
I've see a general convention of putting `private` methods at the bottom of the class

doulikecookiedough replied:
Noted, I'll keep this in mind moving forward. Thank you!

artntek commented on test/edu/ucsb/nceas/metacat/dataone/D1AuthHelperTest.java:
```diff
@@ -157,44 +200,270 @@ public void setUp() throws Exception {
+            // Get a default rightsHolderGroup list
+            List<Group> testRightsHolderGroup = getGenericRightsHolderGroup(null);
+
+            SubjectInfo mockSInfo = Mockito.mock(SubjectInfo.class);
+            when(mockSInfo.getGroupList()).thenReturn(testRightsHolderGroup);
+
+            CNode mockCN = Mockito.mock(CNode.class);
+            // .listSubjects(...) makes a network call
+            when(mockCN.listSubjects(eq(null), any(), eq(null), anyInt(), anyInt()))
+                .thenReturn(mockSInfo);
```
why can't we move these lines inside the private method, too? They are repeated in all 4 tests

doulikecookiedough replied:
I left them out as I had wanted the tests to show more of the mocking - but you're right, they are repeated unnecessarily (and we can always look at the new method `getMockCN` to see how the mocking works). This has been refactored, thank you!

doulikecookiedough commented on test/edu/ucsb/nceas/metacat/dataone/D1AuthHelperTest.java:
```diff
@@ -237,16 +199,8 @@ public void testExpandRightsHolder() throws Exception {
+            CNode mockCN = getMockCN(null);
 
             // .getCN() makes a network call
             mockD1Client.when(D1Client::getCN).thenReturn(mockCN);
```
Thanks for pointing that out - I have revised and made changes to the `getMockCN` method, and renamed it to `initMockCN`.

doulikecookiedough commented:
Thank you for reviewing this PR @taojing2002!
